documentation/modules/managing/con-consumer-config-properties.adoc

```diff
@@ -0,0 +1,252 @@
+when Kafka receives a fetch request from the consumer it will respond when the first of either threshold is reached.
+
+Conversely, you can adjust the `fetch.max.wait.ms` and `fetch.min.bytes` properties lower to improve end-to-end latency.
+
+----
+# ...
+fetch.max.wait.ms=500 <1>
+fetch.min.bytes=16384 <2>
+# ...
+----
+<1> The maximum time in milliseconds the broker will wait before completing fetch requests.
+The default is `500` milliseconds.
+<2> If a minimum batch size in bytes is used, a request is sent when the minimum is reached, or messages have been queued for longer than `fetch.max.wait.ms` (whichever comes sooner).
+Adding the delay allows batches to accumulate messages up to the batch size.
+
+.Lowering latency by increasing the fetch request size
+
+Use the `fetch.max.bytes` and `max.partition.fetch.bytes` properties to increase the maximum amount of data fetched by the consumer from the Kafka broker.
+
+The `fetch.max.bytes` property sets a maximum limit in bytes on the amount of data fetched from the broker at one time.
+
+The `max.partition.fetch.bytes` sets a maximum limit in bytes on how much data is returned for each partition,
+which must always be larger than the number of bytes set in the broker or topic configuration for `max.message.bytes`.
+
+The maximum amount of memory a client can consume is calculated approximately as:
+
+[source,shell,subs="+quotes,attributes"]
+----
+_NUMBER-OF-BROKERS_ * fetch.max.bytes and _NUMBER-OF-PARTITIONS_ * max.partition.fetch.bytes
+----
+
+If memory usage can accommodate it, you can increase the values of these two properties.
+By allowing more data in each request, latency is improved as there are fewer fetch requests.
+
+----
+# ...
+fetch.max.bytes=52428800 <1>
+max.partition.fetch.bytes=1048576 <2>
+# ...
+----
+<1> The maximum amount of data in bytes returned for a fetch request.
+<2> The maximum amount of data in bytes returned for each partition.
+
+== Avoiding data loss or duplication when committing offsets
+
+The Kafka _auto-commit mechanism_ allows a consumer to commit the offsets of messages automatically.
+If enabled, the consumer will commit offsets received from polling the broker at 5000ms intervals.
+
+The auto-commit mechanism is convenient, but it introduces a risk of data loss and duplication.
+If a consumer has fetched and transformed a number of messages, but the system crashes with processed messages in the consumer buffer when performing an auto-commit, that data is lost.
+If the system crashes after processing the messages, but before performing the auto-commit, the data is duplicated on another consumer instance after rebalancing.
+
+To minimize the likelihood of data loss or duplication, set `enable.auto.commit` to `false` and develop your client application to have more control over committing offsets.
+Or use `auto.commit.interval.ms=1` to decrease the intervals between commits.
```

Comment:
I think automatic offset commits can provide at least once delivery as well. From the apche kafka docs:

> Note: Using automatic offset commits can also give you "at-least-once" delivery, but the requirement is that you must consume all data returned from each call to poll(Duration) before any subsequent calls, or before closing the consumer. If you fail to do either of these, it is possible for the committed offset to get ahead of the consumed position, which results in missing records. The advantage of using manual offset control is that you have direct control over when a record is considered "consumed."

Reference: https://kafka.apache.org/23/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html

Reply:
That is true in some cases ... it really depends on how you process these messages (synchronously before the next `poll()` call or asynchronously).

Reply:
I've expanded a little.

> The auto-commit mechanism is convenient, but it introduces a risk of data loss and duplication. If a consumer has fetched and transformed a number of messages, but the system crashes with processed messages in the consumer buffer when performing an auto-commit, that data is lost. If the system crashes after processing the messages, but before performing the auto-commit, the data is duplicated on another consumer instance after rebalancing.
> 
> Auto-committing can avoid creating duplicates only when all messages are consumed before the next poll to the broker,  or the consumer closes.
> 
> To minimize the likelihood of data loss or duplication, you can set `enable.auto.commit` to `false` and develop your client application to have more control over committing offsets. Or you can use `auto.commit.interval.ms=1` to decrease the intervals between commits.

Reply:
@PaulRMellor Your updated version is wrong. Auto-committing can not avoid creating duplicates but you can avoid data loss.

Reply:
I think this sentence should sound like this:

```
Auto-committing can avoid data loss only when all messages are processed before ...
```